### PR TITLE
JSUI-2571 Add maximumNumberOfExpandedFacets option and enable collapsing by default

### PR DESCRIPTION
**Pull request by ThibodeauJF**
UTs are coming, this is mainly for the doc now
https://coveord.atlassian.net/browse/JSUI-2571

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: unitTests/ui/DynamicFacet/DynamicFacetTest.ts [hunk 1]
```diff
@@ -227,6 +227,16 @@ export function DynamicFacetTest() {
       expect($$(test.cmp.element).hasClass('coveo-dynamic-facet-collapsed')).toBe(true);
     });
 
+    it(`when enableCollapse is false & collapsedByDefault options is true
```
Comment (olamothe) [inline unitTests/ui/DynamicFacet/DynamicFacetTest.ts]: Would be also nice to detect those kind of invalid configuration (options that contradict one another) during component initialization/constructor, and log warnings in the console.

File: src/ui/DynamicFacetManager/DynamicFacetManager.ts [hunk 2]
```diff
@@ -71,7 +72,18 @@ export class DynamicFacetManager extends Component {
      */
     compareFacets: ComponentOptions.buildCustomOption<IDynamicFacetManagerCompareFacet>(() => {
       return null;
-    })
+    }),
+    /**
+     * Defines the maximum number of expanded facets inside the manager.
```
Comment (jpmarceau) [inline src/ui/DynamicFacetManager/DynamicFacetManager.ts]: Suggestion:
Defines the maximum number -> The maximum number

File: src/ui/DynamicFacetManager/DynamicFacetManager.ts [hunk 3]
```diff
@@ -71,7 +72,18 @@ export class DynamicFacetManager extends Component {
...
+     * The rest of the facets will be collapsed.
```
Comment (jpmarceau) [inline src/ui/DynamicFacetManager/DynamicFacetManager.ts]: Suggestion: Remaining facets are collapsed.

File: src/ui/DynamicFacetManager/DynamicFacetManager.ts [hunk 4]
```diff
@@ -71,7 +72,18 @@ export class DynamicFacetManager extends Component {
...
+     *
+     * **Note:**
+     * Will prioritize the facets that possess active values, then will prioritize the first facets.
```
Comment (jpmarceau) [inline src/ui/DynamicFacetManager/DynamicFacetManager.ts]: Suggestion: Prioritizes facets with active values, and then prioritizes first facets.

File: src/ui/DynamicFacetManager/DynamicFacetManager.ts [hunk 5]
```diff
@@ -71,7 +72,18 @@ export class DynamicFacetManager extends Component {
...
+     * If more there are more facets that possess active values than the `maximumNumberOfExpandedFacets` option, it will override it.
```
Comment (jpmarceau) [inline src/ui/DynamicFacetManager/DynamicFacetManager.ts]: If the number of facets with active values exceeds the value of the `maximumNumberOfExpandedFacets` option, it overrides the option.

Comment (ThibodeauJF) [reply src/ui/DynamicFacetManager/DynamicFacetManager.ts]: Wow I made a nice typo 😅 That's why we need you guys!

File: src/ui/DynamicFacet/DynamicFacet.ts [hunk 6]
```diff
@@ -707,6 +709,12 @@ export class DynamicFacet extends Component implements IAutoLayoutAdjustableInsi
     this.updateAppearance();
   }
 
+  private verifyCollapsingConfiguration() {
+    if (this.options.collapsedByDefault && !this.options.enableCollapse) {
+      this.logger.warn('The "collapsedByDefault" option is "true" while the "enableCollapse" is "false"');
```
Comment (olamothe) [inline src/ui/DynamicFacet/DynamicFacet.ts]: As a rule of thumb, it's always a good idea to print the component object + HTMLElement reference that logged the error, so that you can click on it in the console and Chrome, for example, will highlight it in the DOM.

Comment (ThibodeauJF) [reply src/ui/DynamicFacet/DynamicFacet.ts]: Doesn't logger automatically does that? It prints the component before the message and it contains the html element too

File: unitTests/ui/DynamicFacet/DynamicFacetTest.ts (hunk 1 above)
Comment (orobert-coveo) [reply unitTests/ui/DynamicFacet/DynamicFacetTest.ts]: He already does with the function `verifyCollapsingConfiguration`

File: unitTests/ui/DynamicFacetManagerTest.ts [hunk 7]
```diff
@@ -28,6 +28,13 @@ export function DynamicFacetManagerTest() {
         DynamicFacetTestUtils.createAdvancedFakeFacet({ id: 'test2', field: '@test2', numberOfValues: 5 }).cmp,
```
Comment (samisayegh) [inline unitTests/ui/DynamicFacetManagerTest.ts]: Are the tests for the new functionality going to be included in a separate PR?

File: src/ui/DynamicFacet/DynamicFacet.ts (hunk 6 above)
Comment (ThibodeauJF) [reply src/ui/DynamicFacet/DynamicFacet.ts]: you're right I had already managed the collapsedByDefault but now that you mentioned enableCollapse I just refactored things a bit to manage it too

File: unitTests/ui/DynamicFacetManagerTest.ts (hunk 7 above)
Comment (ThibodeauJF) [reply unitTests/ui/DynamicFacetManagerTest.ts]: I'll do it soon on this same PR

File: src/ui/DynamicFacetManager/DynamicFacetManager.ts [hunk 8]
```diff
@@ -71,7 +72,20 @@ export class DynamicFacetManager extends Component {
...
+    }),
+    /**
+     * The maximum number of expanded facets inside the manager.
+     * Remaining facets are collapsed.
+     *
+     * **Note:**
+     * Prioritizes facets with active values, and then prioritizes first facets.
+     * If the number of facets with active values exceeds the value of the `maximumNumberOfExpandedFacets` option, it overrides the option.
+     *
+     * Using the value `-1` will disable the feature and keep all facet expanded.
```
Comment (ThibodeauJF) [inline src/ui/DynamicFacetManager/DynamicFacetManager.ts]: @jpmarceau added this sweet line

Comment (jpmarceau) [reply src/ui/DynamicFacetManager/DynamicFacetManager.ts]: Thanks for letting me know :)
will disable -> disables
facet -> facets

**Comment by ThibodeauJF** [comment]
Rerequested review mostly for UTs change and the commit https://github.com/coveo/search-ui/pull/1165/commits/d4185127ac5200d80f25038f1037fbd1abb03dc5

File: src/ui/DynamicFacet/DynamicFacet.ts (hunk 6 above)
Comment (samisayegh) [reply src/ui/DynamicFacet/DynamicFacet.ts]: The warning right now reports what the state is but it is missing an actionable recommendation/insight. Either it should suggest to set "collapsedByDefault" to `false`, or it should explain that the `collapsedByDefault` option will have no effect.

Also, what about adding the `dependsOn` property to the option so that the IEditor will render it hierarchically?

File: unitTests/ui/DynamicFacet/DynamicFacetTestUtils.ts [hunk 9]
```diff
@@ -8,7 +8,7 @@ import { IFacetResponse } from '../../../src/rest/Facet/FacetResponse';
 export class DynamicFacetTestUtils {
   static allOptions(options?: IDynamicFacetOptions) {
     return {
-      field: '@dummy',
+      field: `@dummy${Math.round(Math.random() * 100000)}`,
```
Comment (samisayegh) [inline unitTests/ui/DynamicFacet/DynamicFacetTestUtils.ts]: If this was to ensure you had different values for the field, I would prefer passing the `options` property to specify the field rather than having a chance no matter how small of the same test passing and failing on successive runs.

File: unitTests/ui/DynamicFacetManagerTest.ts [hunk 10]
```diff
@@ -18,25 +19,22 @@ export function DynamicFacetManagerTest() {
...
       options = {};
       initializeFacets();
-      initializeComponent();
-      test.cmp.ensureDom();
+      initializeManager();
     });
 
     function initializeFacets() {
       facets = [
-        DynamicFacetTestUtils.createAdvancedFakeFacet({ id: 'test1', field: '@test1', numberOfValues: 10 }).cmp,
```
Comment (samisayegh) [inline unitTests/ui/DynamicFacetManagerTest.ts]: Related to the comment above, I'm curious why you made this change?

Comment (ThibodeauJF) [reply unitTests/ui/DynamicFacetManagerTest.ts]: I added the method initializeManyFacets and didn't want to specify a field for each, but I guest I could just use the index as well
